File: core/src/main/java/nl/t64/game/rpg/components/party/InventoryDescription.java
package nl.t64.game.rpg.components.party;

import lombok.Getter;
import nl.t64.game.rpg.constants.ThreeState;


@Getter
public class InventoryDescription {

    final Object key;       // Object can be SuperEnum or String.
    final Object value;     // Object can be Integer or String.
    final ThreeState compare;

    InventoryDescription(Object key, Object value, InventoryItem item, HeroItem hero) {
        this.key = key;
        this.value = value;
        this.compare = isEnough(item, hero);
    }

    InventoryDescription(Object key, Object value, InventoryItem item1, InventoryItem item2) {
        this.key = key;
        this.value = value;
        this.compare = compare(item1, item2);
    }

    private ThreeState isEnough(InventoryItem item, HeroItem hero) {
        if (value.equals(0)) {
            return ThreeState.SAME;
        }
        if (key instanceof InventoryMinimal) {
            return ((InventoryMinimal) key).createMessageIfHeroHasNotEnoughFor(item, hero)
                                           .isEmpty() ? ThreeState.SAME : ThreeState.LESS;
        }
        return ThreeState.SAME;
    }

    private ThreeState compare(InventoryItem item1, InventoryItem item2) {
        if (value instanceof Integer) {
            if (key instanceof StatItemId) {
                return compareStats(item1, item2);
            } else if (key instanceof SkillItemId) {
                return compareSkills(item1, item2);
            } else if (key instanceof CalcAttributeId) {
                return compareCalcs(item1, item2);
            }
        }
        return ThreeState.SAME;
    }

    private ThreeState compareStats(InventoryItem item1, InventoryItem item2) {
        StatItemId statItemId = (StatItemId) key;
        if (item1.getAttributeOfStatItemId(statItemId) < item2.getAttributeOfStatItemId(statItemId)) {
            return ThreeState.LESS;
        } else if (item1.getAttributeOfStatItemId(statItemId) > item2.getAttributeOfStatItemId(statItemId)) {
            return ThreeState.MORE;
        } else {
            return ThreeState.SAME;
        }
    }

    private ThreeState compareSkills(InventoryItem item1, InventoryItem item2) {
        SkillItemId skillItemId = (SkillItemId) key;
        if (item1.getAttributeOfSkillItemId(skillItemId) < item2.getAttributeOfSkillItemId(skillItemId)) {
            return ThreeState.LESS;
        } else if (item1.getAttributeOfSkillItemId(skillItemId) > item2.getAttributeOfSkillItemId(skillItemId)) {
            return ThreeState.MORE;
        } else {
            return ThreeState.SAME;
        }
    }

    private ThreeState compareCalcs(InventoryItem item1, InventoryItem item2) {
        CalcAttributeId calcAttributeId = (CalcAttributeId) key;
        if (calcAttributeId.equals(CalcAttributeId.WEIGHT)) {
            return ThreeState.SAME;
        }
        if (item1.getAttributeOfCalcAttributeId(calcAttributeId) < item2.getAttributeOfCalcAttributeId(calcAttributeId)) {
            return ThreeState.LESS;
        } else if (item1.getAttributeOfCalcAttributeId(calcAttributeId) > item2.getAttributeOfCalcAttributeId(calcAttributeId)) {
            return ThreeState.MORE;
        } else {
            return ThreeState.SAME;
        }
    }

}

File: core/src/nl/t64/game/rpg/components/party/inventory/InventoryDescription.kt
package nl.t64.game.rpg.components.party.inventory

import nl.t64.game.rpg.components.party.CalcAttributeId
import nl.t64.game.rpg.components.party.HeroItem
import nl.t64.game.rpg.components.party.skills.SkillItemId
import nl.t64.game.rpg.components.party.stats.StatItemId


class InventoryDescription {

    val key: Any        // Any can be SuperEnum or String.
    val value: Any      // Any can be Integer or String.
    val compare: ThreeState

    constructor(key: Any, value: Any) {
        this.key = key
        this.value = value
        this.compare = ThreeState.SAME
    }

    constructor(key: Any, value: Any, item: InventoryItem, hero: HeroItem) {
        this.key = key
        this.value = value
        this.compare = isEnough(item, hero)
    }

    constructor(key: Any, value: Any, item1: InventoryItem, item2: InventoryItem) {
        this.key = key
        this.value = value
        this.compare = compare(item1, item2)
    }

    private fun isEnough(item: InventoryItem, hero: HeroItem): ThreeState {
        return when {
            value == 0 -> ThreeState.SAME
            key is InventoryMinimal && key.createMessageIfHeroHasNotEnoughFor(item, hero) != null -> ThreeState.LESS
            else -> ThreeState.SAME
        }
    }

    private fun compare(item1: InventoryItem, item2: InventoryItem): ThreeState {
        return when (value) {
            is Int -> when (key) {
                is StatItemId -> compareStats(item1, item2)
                is SkillItemId -> compareSkills(item1, item2)
                is CalcAttributeId -> compareCalcs(item1, item2)
                else -> ThreeState.SAME
            }
            "0" -> ThreeState.LESS
            else -> ThreeState.SAME
        }
    }

    private fun compareStats(item1: InventoryItem, item2: InventoryItem): ThreeState {
        key as StatItemId
        return when {
            item1.getAttributeOfStatItemId(key) < item2.getAttributeOfStatItemId(key) -> ThreeState.LESS
            item1.getAttributeOfStatItemId(key) > item2.getAttributeOfStatItemId(key) -> ThreeState.MORE
            else -> ThreeState.SAME
        }
    }

    private fun compareSkills(item1: InventoryItem, item2: InventoryItem): ThreeState {
        key as SkillItemId
        return when {
            item1.getAttributeOfSkillItemId(key) < item2.getAttributeOfSkillItemId(key) -> ThreeState.LESS
            item1.getAttributeOfSkillItemId(key) > item2.getAttributeOfSkillItemId(key) -> ThreeState.MORE
            else -> ThreeState.SAME
        }
    }

    private fun compareCalcs(item1: InventoryItem, item2: InventoryItem): ThreeState {
        key as CalcAttributeId
        return when {
            key == CalcAttributeId.WEIGHT -> ThreeState.SAME
            item1.getAttributeOfCalcAttributeId(key) < item2.getAttributeOfCalcAttributeId(key) -> ThreeState.LESS
            item1.getAttributeOfCalcAttributeId(key) > item2.getAttributeOfCalcAttributeId(key) -> ThreeState.MORE
            else -> ThreeState.SAME
        }
    }

}
